cmd/create.go

```diff
@@ -137,104 +202,341 @@ type createConfig struct {
+}
+
+// newInvalidTemplateError creates an error stating that the given template
+// is not available for the given runtime, and a verbose list of valid options.
+// The runtime is expected to already have been validated.
+func newInvalidTemplateError(client *fn.Client, runtime, template string) error {
+	b := strings.Builder{}
+	fmt.Fprintf(&b, "The template '%v' was not found for language runtime '%v'.\n", template, runtime)
+	fmt.Fprintln(&b, "Available templates for this language runtime are:")
+	templates, err := client.Templates().List(runtime)
+	if err != nil {
+		return err
+	}
+	for _, v := range templates {
+		fmt.Fprintf(&b, "  %v\n", v)
+	}
+	return errors.New(b.String())
+}
+
+// Validate the current state of the config, returning any errors.
+// Note this is a deeper validation using a client already configured with a
+// preliminary config object from flags/config, such that the client instance
+// can be used to determine possible values for runtime, templates, etc.  a
+// pre-client validation should not be required, as the Client does its own
+// validation.
+func (c createConfig) Validate(client *fn.Client) (err error) {
+	// Client validates both language runtime and template exist, defaulting
+	// them if not (to 'node' and 'http', respectively).  However, if either of
+	// them are invalid, or the chosen combination does not exist, the error
+	// message is a rather terse one-liner.  This is suitable for libraries, but
+	// for a CLI it behooves us to be more verbose, including valid options for
+	// each.  So here, we check that the values entered (if any) are both valid
+	// and valid together.
+	if c.Runtime != "" && !isValidRuntime(client, c.Runtime) {
+		return newInvalidRuntimeError(client, c.Runtime)
+	}
+
+	if c.Template != "" && !isValidTemplate(client, c.Runtime, c.Template) {
+		return newInvalidTemplateError(client, c.Runtime, c.Template)
+	}
+
+	// TODO: refactor to be git-like with no name at time of creation, but rather
+	// with named deployment targets in a one-to-many configuration.
+	dirName, _ := deriveNameAndAbsolutePathFromPath(c.Path)
+	return utils.ValidateFunctionName(dirName)
+	// NOTE: perhaps additional validation would be of use here in the CLI, but
+	// the client libray itself is ultimately responsible for validating all input
+	// prior to exeuting any requests.
+}
+
+// prompt the user with value of config members, allowing for interactively
+// mutating the values. The provided clientFn is used to construct a transient
+// client for use during prompt autocompletion/suggestions (such as suggesting
+// valid templates)
+func (c createConfig) prompt(client *fn.Client) (createConfig, error) {
+	var qs []*survey.Question
+
+	runtimes, err := client.Runtimes()
+	if err != nil {
+		return createConfig{}, err
+	}
+
+	// First ask for path...
+	qs = []*survey.Question{
 		{
 			Name: "path",
 			Prompt: &survey.Input{
-				Message: "Project path:",
+				Message: "Function Path:",
```

Comment:
Saving "project" for when, in the near future, we create some kind of project which aggregates multiple Functions into a single project.  It seems appropriate to stick with just "Function" here.